src/serde/de_tree.rs

```diff
@@ -0,0 +1,327 @@
+use std::convert::TryInto;
+use std::io::{Error, Read, Result, Write};
+
+use sha2::Digest;
+
+use crate::sha2::Sha256;
+
+use super::parse_atom::decode_size_with_offset;
+use super::utils::{copy_exactly, skip_bytes};
+
+const MAX_SINGLE_BYTE: u8 = 0x7f;
+const CONS_BOX_MARKER: u8 = 0xff;
+
+struct ShaWrapper(Sha256);
+
+impl Write for ShaWrapper {
+    fn write(&mut self, blob: &[u8]) -> std::result::Result<usize, Error> {
+        self.0.update(blob);
+        Ok(blob.len())
+    }
+    fn flush(&mut self) -> std::result::Result<(), Error> {
+        Ok(())
+    }
+}
+
+/// This data structure is used with `parse_triples`, which returns a triple of
+/// integer values for each clvm object in a tree.
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum ParsedTriple {
+    Atom {
+        start: u64,
+        end: u64,
+        atom_offset: u32,
+    },
+    Pair {
+        start: u64,
+        end: u64,
+        right_index: u32,
+    },
+}
+
+enum ParseOpRef {
+    ParseObj,
+    SaveCursor(usize),
+    SaveIndex(usize),
+}
+
+fn sha_blobs(blobs: &[&[u8]]) -> [u8; 32] {
+    let mut h = Sha256::new();
+    for blob in blobs {
+        h.update(blob);
+    }
+    h.finalize()
+        .as_slice()
+        .try_into()
+        .expect("wrong slice length")
+}
+
+fn tree_hash_for_byte(b: u8, calculate_tree_hashes: bool) -> Option<[u8; 32]> {
+    if calculate_tree_hashes {
+        Some(sha_blobs(&[&[1, b]]))
+    } else {
+        None
+    }
+}
+
+fn skip_or_sha_bytes<R: Read>(
+    f: &mut R,
+    skip_size: u64,
+    calculate_tree_hashes: bool,
+) -> Result<Option<[u8; 32]>> {
+    if calculate_tree_hashes {
+        let mut h = Sha256::new();
+        h.update([1]);
+        let mut w = ShaWrapper(h);
+        copy_exactly(f, &mut w, skip_size)?;
+        let r: [u8; 32] =
+            w.0.finalize()
+                .as_slice()
+                .try_into()
+                .expect("wrong slice length");
+        Ok(Some(r))
+    } else {
+        skip_bytes(f, skip_size)?;
+        Ok(None)
+    }
+}
+
+/// parse a serialized clvm object tree to an array of `ParsedTriple` objects
+
+/// This alternative mechanism of deserialization generates an array of
+/// references to each clvm object. A reference contains three values:
+/// a start offset within the blob, an end offset, and a third value that
+/// is either: an atom offset (relative to the start offset) where the atom
+/// data starts (and continues to the end offset); or an index in the array
+/// corresponding to the "right" element of the pair (in which case, the
+/// "left" element corresponds to the current index + 1).
+///
+/// Since these values are offsets into the original buffer, that buffer needs
+/// to be kept around to get the original atoms.
+
+type ParsedTriplesOutput = (Vec<ParsedTriple>, Option<Vec<[u8; 32]>>);
```

Comment:
I considered a python version of `CLVMTree` that doesn't keep the entire tree of objects around, but just the top one, and re-parses on demand when `.pair` is invoked, and started writing it before deciding it may be too much new stuff when it's not even clear we need it. If memory turns out to be an issue, I can resurrect that code later.

In this implementation, the hashes (and triples) would be discard except the first after returning to python. Sure we have the whole stack of hashes, but we need to calculate them anyway to get the top node's hash, so it just ends up acting like a "bump allocator" (like how `Allocator`'s `Vec` works).

Reply:
The time to hash after parsing is significantly slower than parsing and tree hashing all at once.